sdk/apimanagement/api-management-custom-widgets-tools/src/utils.ts

```diff
@@ -0,0 +1,170 @@
+ */
+export type TEnvironment = "development" | "publishing" | "runtime" | "error";
+
+/** Information about the widget instance received from the Dev Portal */
+export interface TPortalData {
+  /** web content's origin (URL) of your Dev Portal */
+  origin: string;
+  /** current runtime environment */
+  environment: TEnvironment;
+  /** ID of this particular instance of the widget */
+  instanceId: string;
+}
+
+/** JSON object with all the data you'll receive from the Dev Portal */
+export interface TEditorData<TValues extends TValuesCommon> extends TPortalData {
+  /** values you've set in the admin editor window */
+  values: TValues;
+}
+
+export function getEditorDataPure<TValues extends TValuesCommon>(
+  valuesDefault: TValues,
+  urlSearchParams: URLSearchParams
+): TEditorData<TValues> {
+  try {
+    const urlEditorParams = JSON.parse(
+      decodeURIComponent(urlSearchParams.get(EDITOR_DATA_KEY) ?? "")
+    );
+
+    if (!("origin" in urlEditorParams)) {
+      console.error(
+        "Could not get 'origin' from the search params of the URL:\n" + self.location.href
+      );
+    }
+    return { ...urlEditorParams, values: { ...valuesDefault, ...urlEditorParams.values } };
+  } catch (e) {
+    console.error(
+      `Could not get '${EDITOR_DATA_KEY}' from the search params of the URL:\n` + self.location,
+      e
+    );
+    return { values: valuesDefault, environment: "error", origin: "error", instanceId: "error" };
+  }
+}
+
+/**
+ * Function to get all editor data
+ *
+ * @param valuesDefault - object with your default values to use, just import valuesDefault object from values.ts folder
+ */
+export function getEditorData<TValues extends TValuesCommon>(
+  valuesDefault: TValues
+): TEditorData<TValues> {
+  return getEditorDataPure(valuesDefault, new URLSearchParams(self.location.search));
+}
+
+export function getEditorValuesPure<TValues extends TValuesCommon>(
+  valuesDefault: TValues,
+  urlSearchParams: URLSearchParams
+): TValues {
+  return getEditorDataPure(valuesDefault, urlSearchParams).values;
+}
+
+/**
+ * Function to get values you've set in the admin editor window.
+ *
+ * @param valuesDefault - object with your default values to use, just import valuesDefault object from values.ts folder
+ */
+export function getEditorValues<TValues extends TValuesCommon>(valuesDefault: TValues): TValues {
+  return getEditorValuesPure(valuesDefault, new URLSearchParams(self.location.search));
+}
+
+/**
+ * Type of the onChange function.
+ */
+export type TOnChange<TValues extends TValuesCommon> = (values: Partial<TValues>) => void;
+
+/**
+ * The onChange function itself with 'origin' provided as a param.
+ *
+ * @param origin - web content's origin (URL) of your Dev Portal to send changes to
+ * @param instanceId - ID of this particular instance of the widget
+ * @param values - values that changed
+ */
+export function onChangeWithOrigin<TValues extends TValuesCommon>(
+  origin: TEditorData<TValues>["origin"],
```

Comment:
`originUrl: string` there is no benefit in doing this Type calculations to get string and it is much more readable